language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -32,132 +37,69 @@ mod value_tests;
+/// For instance, `[Value]` is a wrapper of `ValueImpl` that does not allow `Invalid` to be
+/// represented. In that sense instances of `[Value]` are what gets pushed/popped on the
+/// stack and effectively the values representable in Move.
+#[derive(PartialEq, Eq, Debug, Clone)]
+enum ValueImpl {
+    /// Locals are invalid on entry of a function and when moved out.
+    Invalid,
+    // Primitive types
     U64(u64),
+    Address(AccountAddress),
     Bool(bool),
-    String(VMString),
-    Struct(Vec<MutVal>),
     ByteArray(ByteArray),
-}
-
-impl Value {
-    fn size(&self) -> AbstractMemorySize<GasCarrier> {
-        match self {
-            Value::U64(_) | Value::Bool(_) => *CONST_SIZE,
-            Value::Address(_) => AbstractMemorySize::new(ADDRESS_LENGTH as u64),
-            // Possible debate topic: Should we charge based upon the size of the string.
-            // At this moment, we take the view that you should be charged as though you are
-            // copying the string onto the stack here. This doesn't replicate
-            // the semantics that we utilize currently, but this string may
-            // need to be copied at some later time, so we need to charge based
-            // upon the size of the memory that will possibly need to be accessed.
-            Value::String(s) => words_in(AbstractMemorySize::new(s.len() as u64)),
-            Value::Struct(vals) => vals
-                .iter()
-                .fold(*STRUCT_SIZE, |acc, vl| acc.map2(vl.size(), Add::add)),
-            Value::ByteArray(key) => AbstractMemorySize::new(key.len() as u64),
-        }
-    }
+    String(VMString),
 
-    /// Normal code should always know what type this value has. This is made available only for
-    /// tests.
-    #[allow(non_snake_case)]
-    #[doc(hidden)]
-    pub fn to_struct_def_FOR_TESTING(&self) -> StructDef {
-        let values = match self {
-            Value::Struct(values) => values,
-            _ => panic!("Value must be a struct {:?}", self),
-        };
+    /// A struct in Move.
+    Struct(Struct),
 
-        let fields = values
-            .iter()
-            .map(|mut_val| {
-                let val = &*mut_val.peek();
-                match val {
-                    Value::Bool(_) => Type::Bool,
-                    Value::Address(_) => Type::Address,
-                    Value::U64(_) => Type::U64,
-                    Value::String(_) => Type::String,
-                    Value::ByteArray(_) => Type::ByteArray,
-                    Value::Struct(_) => Type::Struct(val.to_struct_def_FOR_TESTING()),
-                }
-            })
-            .collect();
-        StructDef::new(fields)
-    }
+    /// Reference to a local.
+    Reference(Reference),
+    /// Global reference into storage.
+    GlobalRef(GlobalRef),
+    /// A reference to a local.
+    ///
+    /// This value is used to promote a value into a reference lazily when borrowing a local.
+    LocalRef(Reference),
```

Comment:
Later when I was reading the code implementing some API in ValueImpl, I saw something special being done for LocalRef variant.  By that time, I had already forgotten the comment above the declaration of LocalRef variant and I thought that LocalRef is actually a Move reference to a local (which happens to be the Reference variant).  To avoid this kind of confusion, consider renaming LocalRef to PromotedReference, which sets it nicely alongside  Reference.

Reply:
sounds like a better name to me, I'll change it